lib/edgehog_device_forwarder/tokens.ex

```diff
@@ -0,0 +1,67 @@
+# Copyright 2023 SECO Mind Srl
+# SPDX-License-Identifier: Apache-2.0
+
+defmodule EdgehogDeviceForwarder.Tokens do
+  alias EdgehogDeviceForwarder.Tokens.Data
+  alias EdgehogDeviceForwarder.Monitor
+  require Logger
+
+  @tokens EdgehogDeviceForwarder.Cache.for(__MODULE__)
+  @type token :: String.t()
+  @type device_id :: Data.device_id()
+  @type realm :: Data.realm()
+  @type socket :: Data.socket()
+
+  @spec add_device_info(token, device_id, realm) ::
+          :ok | {:error, :token_not_found}
+  def add_device_info(token, device_id, realm) do
+    ConCache.update(@tokens, token, fn
+      nil -> {:error, :token_not_found}
+      token_data -> {:ok, Data.put_device_info(token_data, device_id, realm)}
+    end)
+  end
+
+  @spec reserve(token) :: :ok | {:error, :token_already_exists}
+  def reserve(token) do
+    with {:error, :already_exists} <- ConCache.insert_new(@tokens, token, :reserved) do
+      {:error, :token_already_exists}
+    end
+  end
+
+  @spec register(token, socket) :: :ok | {:error, :invalid_data}
+  def register(token, socket) do
+    ConCache.update(@tokens, token, fn
+      :reserved ->
+        start_monitor(socket, token)
+        {:ok, %Data{socket: socket}}
+
+      other ->
+        "Tokens.Data.register: token #{inspect(token)}: found #{inspect(other)}"
+        |> Logger.notice(tag: "token_data_error")
```

Comment:
Prefer using a function call rather than a pipe with just one function. You could either do something like
```elixir
log_notice = "Tokens.Data.register: token #{inspect(token)}: found #{inspect(other)}"
Logger.notice(log_notice, tag: "token_data_error")
```
or
```elixir
Logger.notice("Tokens.Data.register: token #{inspect(token)}: found #{inspect(other)}", tag: "token_data_error")
```

Reply:
I generally agree with this statement, but in log-specific contexts I find both the message and the tag way more readable this way, as log messages [can be quite long](https://github.com/edgehog-device-manager/edgehog_device_forwarder/pull/8/commits/c064189b15c8377bc5af44b72d7ec92d4f5bc3f9#diff-550264c2983517c8e7e8c50cc0cc2657ef4815c1c1dee72196d73054eaba14d9R100-R104)

I can go on with this change, but imho it's nicer as it is now